### ISSUE TITLE
How can I get the real-time validation status?

### ISSUE THREAD

**Issue by liyiming22** [josdejong/jsoneditor#1072]
How can I get the real-time validation status? (Error or No Errors)
For example: I set a button beside the jsoneditor, and when users input the wrong format json data, the button should be disabled, and when users fix the format issue, the button should be enabled.

I have seen the API documents, but I didn't find the apis like (getValidationStatus or status eg.)

**Comment by josdejong** [comment]
Is `onValidationError` what you mean? It's called when there is a change in the errors (also from an error to no error)

Docs: https://github.com/josdejong/jsoneditor/blob/develop/docs/api.md

**Comment by ChrisAcrobat** [comment]
For me, no. I don't think so.
I use `onChange` to listening for changes, and some parts can't be executed without knowing the current status. Including schema, so a simple `try{JSON.parse}` would not be sufficient.
It would be possible if it is promised that `onValidationError` is executed before `onChange`, but I can't find any confirmation abort that in the documentation.
So a `{function} getValidationErrors(errors)` would be handy.

Just for reference, in Ace you can get the information on any change with this:
``` JavaScript
aceEditor.getSession().on('changeAnnotation', ()=>{
	console.log(_editor.getSession().getAnnotations().filter(a=>a.type==='error'));
});
```

**Comment by ChrisAcrobat** [comment]
After a quick look around, I think it would be sufficient to do `function getValidationErrors(){return this.lastSchemaErrors}` in `textmode.js` and `treemode.js`. I will look in to it more next time I have time.

**Comment by josdejong** [comment]
👍

**Comment by ChrisAcrobat** [comment]
After a bit more digging (way to much, because I missed the obvious 😅) I found that `this.lastSchemaErrors` is already available with `jsonEditor.lastSchemaErrors`. 🤦‍♂️
But as I feared, it seams to always be updated after `onChange` has been called.
The investigation continues. 🔎

**Comment by josdejong** [comment]
Just an other idea: you can always create your own instance of Ajc with the same JSON schema and execute that at the moment you want with the current JSON.

**Comment by ChrisAcrobat** [comment]
Yeah, I thought of that as well. I but at the same time, I thought that it should already be include somehow and did't want to reinvent the wheel.
I think I have found the problem. `onChange` is always executed before, because the validation is denounced. Returning a promise from the denounce should do the trick.

**Comment by ChrisAcrobat** [comment]
Fixed it! At least for `textmode.js`. Now `onValidationError` will always be executed first, then `onChange`. I will not try `jsonEditor.lastSchemaErrors` any future, because it was only set if `typeof this.options.onValidationError === 'function'` (at least in `textmode.js`). But it should not be a big deal to just move `this.lastSchemaErrors = parseErrors` outside of the if-statment and create the suggested `function getValidationErrors(){return this.lastSchemaErrors}`. But I don't currently have time to investigate the other modes and make sure that `jsonEditor.lastSchemaErrors` is not hindered by any other `typeof this.options.onValidationError === 'function'` (at least in `textmode.js`-like checks. I will instead use `onValidationError` to update my error variable and then use it in `onChange`. Like I suggested [previosly](https://github.com/josdejong/jsoneditor/issues/1072#issuecomment-830696950).
I will make a PR after I have tested it some more.

**Comment by ChrisAcrobat** [comment]
Update. I will probably take the time to implement `function getValidationErrors(){return this.lastSchemaErrors}`, since it will make my code more clean and I will probably not be able to drop it until I have solved it. 🙃

**Comment by ChrisAcrobat** [comment]
Update.
I have solved the feature (df59de6e87feba062de748e489039bdb5f6c152c). I now only have to update the documentation, then I'll make a PR.

**Comment by ChrisAcrobat** [comment]
@josdejong hope you like it! 👍

**Comment by ChrisAcrobat** [comment]
I changed back #1313 into a draft again, I noticed that the minified version generates a bug not present in the _raw_ version.
Investigating and will reply when I have figured it out.

**Comment by ChrisAcrobat** [comment]
I didn't run `npm run build`. 🤦‍♂️
Everything is fine and working, I'll republish the PR soon.

**Comment by josdejong** [comment]
Published now in v9.5.0, fixed via #1319
